### PR TITLE
Avoid all local git repository actions if rebasing through the GitLab API

### PR DESCRIPTION
**Pull request by slovdahl**
Co-authored-by: Con <con.symeonides@mindfoundry.ai>

### CODE REVIEW

File: marge/app.py [hunk 1]
```diff
@@ -310,7 +310,10 @@ def main(args=None):
             options.ci_timeout = timedelta(minutes=options.max_ci_time_in_minutes)
 
         if options.batch:
-            logging.warning('Experimental batch mode enabled')
+            if options.rebase_remotely:
+                raise Exception("Batch processing does not work together with rebase remotely.")
+            else:
+                logging.warning('Experimental batch mode enabled')
```
Comment (slovdahl) [inline marge/app.py]: There might be more cases where `rebase_remotely` does not work. What about everything related to Marge modifying the commit message? :thinking:

File: marge/single_merge_job.py [hunk 2]
```diff
@@ -49,46 +51,57 @@ def update_merge_request_and_accept(self, approvals):
...
+                    self._api,
                 )
-            except GitLabRebaseResultMismatch as err:
-                if first_iteration:
-                    log.info(
-                        "Gitlab rebase didn't give expected result."
-                        "This is expected immediately after rebase. Retrying."
+
+                if source_branch.commit.id != merge_request.diff_refs.base_sha:
+                    self.synchronize_using_gitlab_rebase(merge_request)
```
Comment (slovdahl) [inline marge/single_merge_job.py]: Here's the most important fix. Instead of always going through the `self.fetch_source_project`, `self.get_target_project`, `self.update_from_target_branch_and_push` code paths, we just check if the SHA of the target branch HEAD and the base SHA of the MR matches, otherwise we assume the MR needs to be rebased. The `self.update_from_target_branch_and_push` call requires a local git repo checkout.

This reminds me that we should probably do a final check that the target branch HEAD SHA and MR base SHA matches after CI has passed too.

File: marge/store.py [hunk 3]
```diff
@@ -40,12 +40,13 @@ def repo_for_project(self, project):
...
             if not self._skip_clone:
                 repo.clone()
-            repo.config_user_info(
-                user_email=self._user.email,
-                user_name=self._user.name,
-            )
+                repo.config_user_info(
+                    user_email=self._user.email,
+                    user_name=self._user.name,
+                )
```
Comment (slovdahl) [inline marge/store.py]: This is important too. We cannot set the git config if we don't have any repository cloned.

**Comment by slovdahl** [comment]
Trying to fix this error after #5 was merged:

```
2022-12-22 06:28:28,336 INFO Running git -C /tmp_mbpkyvd/tmpjx6ptazt config user.email marge-bot@hibox.tv
2022-12-22 06:28:28,354 WARNING git returned 128
2022-12-22 06:28:28,355 WARNING stdout: b''
2022-12-22 06:28:28,355 WARNING stderr: b'fatal: not in a git directory\n'
2022-12-22 06:28:28,355 ERROR Couldn't initialize repository for project!
Traceback (most recent call last):
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/git.py", line 185, in git
    return _run(*command, env=env, check=True, timeout=timeout_seconds)
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/git.py", line 211, in _run
    retcode, process.args, output=stdout, stderr=stderr,
subprocess.CalledProcessError: Command '[b'git', b'-C', b'/tmp_mbpkyvd/tmpjx6ptazt', b'config', b'user.email', b'marge-bot@hibox.tv']' returned non-zero exit status 128.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/bot.py", line 158, in _process_merge_requests
    repo = repo_manager.repo_for_project(project)
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/store.py", line 47, in repo_for_project
    user_name=self._user.name,
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/git.py", line 40, in config_user_info
    self.git('config', 'user.email', user_email)
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/git.py", line 190, in git
    raise GitError(err) from err
marge.git.GitError: Command '[b'git', b'-C', b'/tmp_mbpkyvd/tmpjx6ptazt', b'config', b'user.email', b'marge-bot@hibox.tv']' returned non-zero exit status 128.
Exception occured
Traceback (most recent call last):
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/git.py", line 185, in git
    return _run(*command, env=env, check=True, timeout=timeout_seconds)
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/git.py", line 211, in _run
    retcode, process.args, output=stdout, stderr=stderr,
subprocess.CalledProcessError: Command '[b'git', b'-C', b'/tmp_mbpkyvd/tmpjx6ptazt', b'config', b'user.email', b'marge-bot@hibox.tv']' returned non-zero exit status 128.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/bin/.marge.app-wrapped", line 6, in <module>
    marge.app.main()
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/app.py", line 358, in main
    marge_bot.start()
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/bot.py", line 54, in start
    self._run(repo_manager)
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/bot.py", line 72, in _run
    projects,
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/bot.py", line 113, in _process_projects
    self._process_merge_requests(repo_manager, project, merge_requests)
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/bot.py", line 158, in _process_merge_requests
    repo = repo_manager.repo_for_project(project)
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/store.py", line 47, in repo_for_project
    user_name=self._user.name,
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/git.py", line 40, in config_user_info
    self.git('config', 'user.email', user_email)
  File "/nix/store/sddmhcbs3p23jya97rpsv5izm9nv2r3l-python3.6-marge-0.11.0/lib/python3.6/site-packages/marge/git.py", line 190, in git
    raise GitError(err) from err
marge.git.GitError: Command '[b'git', b'-C', b'/tmp_mbpkyvd/tmpjx6ptazt', b'config', b'user.email', b'marge-bot@hibox.tv']' returned non-zero exit status 128.
```

**Comment by slovdahl** [comment]
cc @csymeonides-mf, if you are interested.

File: marge/single_merge_job.py (hunk 2 above)
Comment (slovdahl) [reply marge/single_merge_job.py]: Ah, but that is not needed :thinking: It's already handled.

image

File: marge/app.py (hunk 1 above)
Comment (perlun) [reply marge/app.py]: True... That will very likely not work with the current implementation. :thinking: But you **could** actually set the `merge_commit_message` when merging using the API as well: https://docs.gitlab.com/ee/api/merge_requests.html#merge-a-merge-request

File: marge/job.py [hunk 4]
```diff
@@ -153,7 +154,9 @@ def get_mr_ci_status(self, merge_request, commit_sha=None):
                 merge_request.source_branch,
                 self._api,
             )
+
         current_pipeline = next(iter(pipeline for pipeline in pipelines if pipeline.sha == commit_sha), None)
+        log.debug('Current pipeline: %s', current_pipeline)
```
Comment (perlun) [inline marge/job.py:159]: Should this remain or be reverted?

File: marge/merge_request.py [hunk 5]
```diff
@@ -183,6 +188,7 @@ def rebase(self):
         self.refetch_info()
 
         if not self.rebase_in_progress:
+            log.debug('Rebasing through GitLab API..')
```
Comment (perlun) [inline marge/merge_request.py:191]: Likewise, do we want to keep the more verbose logging?

File: marge/merge_request.py [hunk 6]
```diff
@@ -198,6 +204,12 @@ def rebase(self):
             if not self.rebase_in_progress:
                 if self.merge_error:
                     raise MergeRequestRebaseFailed(self.merge_error)
+
+                # GitLab does not seem to always return the latest SHA even after
+                # rebase_in_progress starts returning false, hence we need to wait
+                # a little bit more before returning.
+                time.sleep(5)
+                self.refetch_info()
```
Comment (perlun) [inline marge/merge_request.py:212]: :+1: Good that you noted this down. It's a bit of a kludge that we have to do this though, and ideally I would like us to formulate an issue towards GitLab re. this.

File: marge/single_merge_job.py [hunk 7]
```diff
@@ -49,46 +51,71 @@ def update_merge_request_and_accept(self, approvals):
...
+                    _updated_sha
+                )
+                time.sleep(5)
+
+                sha_now = Commit.last_on_branch(source_project.id, merge_request.source_branch, api).id
+                # Make sure no-one managed to race and push to the branch in the
+                # meantime, because we're about to impersonate the approvers, and
+                # we don't want to approve unreviewed commits
+                if sha_now != actual_sha:
+                    raise CannotMerge('Someone pushed to branch while we were trying to merge')
```
Comment (perlun) [inline marge/single_merge_job.py:118]: This hunk is hard to follow along with. :thinking: What is the change going on here?

File: tests/test_single_job.py [hunk 8]
```diff
@@ -205,6 +207,11 @@ def add_reviewers(self, request):
 
     @pytest.fixture()
     def options_factory(self, fusion, add_tested, add_reviewers, add_part_of):
+        if fusion is Fusion.gitlab_rebase and (
+            add_tested or add_reviewers or add_part_of
+        ):
+            pytest.skip('Modifying commit message when using GitLab API rebase is unsupported')
+
```
Comment (perlun) [inline tests/test_single_job.py:229]: The indentation here looks really weird, until  you realize that sad guy on line 212. `):` :joy: Perhaps better to just merge the whole `if` condition on a single line? Right now, it feels like the `pytest.skip()` is flying around there on its own, disconnected from where it "should" be...

File: marge/app.py (hunk 1 above)
Comment (slovdahl) [reply marge/app.py]: Ah, right. Doable but a completely new feature :grin:

**Comment by perlun** [comment]
(Hmm, I found out that I had a bunch of unsubmitted feedback on this one. :see_no_evil:)

File: marge/job.py (hunk 4 above)
Comment (slovdahl) [reply marge/job.py:159]: It's not logged by default (`debug`), so maybe we can keep it?

File: marge/merge_request.py (hunk 5 above)
Comment (slovdahl) [reply marge/merge_request.py:191]: It's not logged by default (`debug`), so maybe we can keep it?

File: marge/single_merge_job.py (hunk 7 above)
Comment (slovdahl) [reply marge/single_merge_job.py:118]: Inside the `if` block is where most of the new logic are this PR (after `if self._config.use_only_gitlab_api and self._options.fusion is Fusion.gitlab_rebase:` that is).

The `else` is just the old code indented one step.

File: tests/test_single_job.py (hunk 8 above)
Comment (slovdahl) [reply tests/test_single_job.py:229]: I think this way of formatting is quite common in the Python world. Not sure if 

```python
    @pytest.fixture()
    def options_factory(self, fusion, add_tested, add_reviewers, add_part_of):
        if fusion is Fusion.gitlab_rebase and (add_tested or add_reviewers or add_part_of):
            pytest.skip('Modifying commit message when using GitLab API rebase is unsupported')
```

is any better in terms of readability of the `if` line.

Comment (perlun) [reply tests/test_single_job.py:229]: I think my main problem with it is that these expressions/statements:

- `add_tested or add_reviewers or add_part_of`
- `pytest.skip('...')`

...are indented on the same level, making them appear to be related somehow, when they are in fact not related at all.

As for whether this is common in the Python world, I can't tell - you are more familiar with that world than me. :slightly_smiling_face: I do think your posted snippet is more readable in fact, at least in terms of "overall readability". The `if` line is also quite fine because of the parentheses.

So I would suggest to go with that approach. :+1: If you insist, I can probably live with the "sad guy"-style indentation though. :grin:

**Review by perlun — APPROVED** [review]

File: tests/test_single_job.py (hunk 8 above)
Comment (slovdahl) [reply tests/test_single_job.py:229]:
```python
        if fusion is Fusion.gitlab_rebase and (add_tested or add_reviewers or add_part_of):
            pytest.skip('Modifying commit message when using GitLab API rebase is unsupported')
```

it is :smiling_imp:
